alternator: improve, document and test table/index name lengths

nyh opened the pull request:
Whereas DynamoDB limits the names of tables, LSIs and GSIs to 255 characters each, Alternator currently has different (and lower) limitations:
 1. A table name must be up to 222 characters.
 2. For a GSI, the sum of the table's and GSI's name length, plus 1, must be up to 222 characters.
 3. For an LSI, the sum of the table's and LSI's name length, plus 2, must be up to 222 characters.

The first patch documents these existing limitations, improves their testing, and fixes a tiny bug found by one of the tests (where UpdateTable adding a GSI's limit testing is off by one).

The second patch unfortunately shows with a reproducer (issue #24598) this limit of 222 is problematic and we may need to lower it: If a user creates a table of length 222 and then enables Alternator streams, Scylla shuts down on an IO error. This will need to be fixed later, but at least this patch properly documents the existing behavior.

No need to backport this patch - it is a very minor improvement that it is unlikely users care about and there is no potential for harm.

Annamikhlin commented:
> @scylladb/releng 15 hours later, the CI didn't complete here. I guess there's some problem.

Yes, basically the CI passed all the relevant steps - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17845/
but finished with some internal error
`21:11:00  hudson.remoting.ProxyException: java.nio.file.FileAlreadyExistsException: /var/lib/jenkins/jobs/scylla-master/jobs/scylla-ci/builds/17845/archive/allure-report.zip`
seems an issue in pkg - checking..

ScyllaPiotr commented on test/alternator/test_gsi_updatetable.py:
```diff
@@ -727,3 +727,41 @@ def test_gsi_backfill_key_limits(dynamodb):
+
+def test_gsi_updatetable_very_long_name_256(table1):
+    with pytest.raises(ClientError, match='ValidationException'):
+        add_and_delete_gsi(table1, 'n' * 255)
```
`256`, not `255`

ScyllaPiotr commented on test/alternator/test_gsi_updatetable.py:
```diff
@@ -727,3 +727,41 @@ def test_gsi_backfill_key_limits(dynamodb):
+def test_gsi_updatetable_very_long_name_222(table1, scylla_only):
+    # If we subtract from 222 the table's name length and an extra 1,
+    # this is how long the GSI's name may be:
+    max = 222 - len(table1.name) - 1
```
Let's call `max` `max_gsi_name_len` instead.

nyh replied:
See my answer above about "max"

ScyllaPiotr commented on test/alternator/test_lsi.py:
```diff
@@ -611,3 +611,40 @@ def test_lsi_missing_attribute(test_table_lsi_1):
+    # If we subtract from 222 the table's name length (we assume that
+    # unique_table_name() always returns the same length) and an extra 2,
+    # this is how long the LSI's name may be:
+    max = 222 - len(unique_table_name()) - 2
```
Let's call `max` `max_lsi_name_len` instead.

nyh replied:
ditto

ScyllaPiotr commented on test/alternator/test_table.py:
```diff
@@ -97,15 +97,14 @@ def test_create_and_delete_table_non_scylla_name(dynamodb):
+    create_and_delete_table(dynamodb, 'n' * 255)
+def test_create_and_delete_table_256(dynamodb):
+    with pytest.raises(ClientError, match='ValidationException'):
+       create_table(dynamodb, 'n' * 256)
```
Why not `create_and_delete_table` here too, especially that the top-level function name promises it? I know this will raise, but suppose it does not raise for some reason.

ScyllaPiotr reviewed (changes requested):
Congrats on the bug with enabling streams!

I'm requesting changes because there is a tiny bug here (or I'm wrong).
Other than that, only small suggestions from my side.

nyh commented on alternator/executor.cc:
```diff
@@ -294,16 +294,16 @@ static std::string view_name(std::string_view table_name, std::string_view index
     }
     std::string ret = std::string(table_name) + delim + std::string(index_name);
-    if (ret.length() > max_table_name_length) {
+    if (ret.length() > max_table_name_length && validate) {
```
I sort-of deliberately chose this order, because validate is almost always true (previously validate=true) and the first condition is almost always false, so it's a tiny tiny tiny bit faster to notice immediately that the first condition is false instead of checking first that validate==true and only then checking the condition.

nyh commented on alternator/executor.cc:
```diff
@@ -1828,7 +1828,7 @@ future<executor::request_return_type> executor::update_table(client_state& clien
                                 "GSI {} already exists in table {}", index_name, table_name));
                         }
-                        if (p.local().data_dictionary().has_schema(keyspace_name, lsi_name(table_name, index_name))) {
+                        if (p.local().data_dictionary().has_schema(keyspace_name, lsi_name(table_name, index_name, false))) {
```
The code adds a GSI (you can only add a GSI in UpdateTable) so it already validates the validity of the GSI name. It was never the intention of this code to check if it's a valid **LSI** name and report an error about a bad LSI name. The only intention was to try to check if an LSI of this name exists (and if it does, report an error). So we want to look up what an LSI with that name have as its name, without validation.

Note that whether or not the lsi_name() function validates valid characters is not important - we already validated them in gsi_name() earlier. The only place this change ever matter is when the length of the table name plus index name plus 1 is exactly 222 - because this is a valid GSI name but not a valid LSI name (because of the silly reason that we used two characters instead of one as a separator for LSIs).

ScyllaPiotr replied:
That's clear, thanks.
Although it would me more prudent to design the `lsi_name()` with a parameter `bool validate` (the name you had there before) that would enable/disable all validations, not only the length - ultimately that's what is aimed here - _just gimme name, no validation!_

nyh commented on test/alternator/test_gsi.py:
```diff
@@ -1474,12 +1474,25 @@ def test_gsi_non_scylla_name(dynamodb):
+    # If we subtract from 222 the table's name length (we assume that
+    # unique_table_name() always returns the same length) and an extra 1,
+    # this is how long the GSI's name may be:
+    max = 222 - len(unique_table_name()) - 1
```
I think "max_gsi_name_len" sounds too general - it sounds like it might be the maximum allowed gsi name length for all GSIs, which isn't true - this is only the maximum for GSIs of this specific table (the formula subtract this table's length from it).
In any case, I think that after a 3-line comment explaining what "max" is, a shorter name is fine. The comment explains much better than a name of any length could have explained.

ScyllaPiotr replied:
> "max_gsi_name_len" sounds too general - it sounds like it might be the maximum allowed gsi name length for all GSIs

I don't agree, the variable is local, so is its context. And it's better than the general `max`.

nyh commented:
Pushed a new version addressing some of the review comments (and answering others in the review). Also I hope that the new version will finally get a working CI.

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -285,7 +285,7 @@ static void validate_table_name(const std::string& name) {
 // The view_name() function assumes the table_name has already been validated
 // but validates the legality of index_name and the combination of both.
-static std::string view_name(std::string_view table_name, std::string_view index_name, const std::string& delim = ":") {
+static std::string view_name(std::string_view table_name, std::string_view index_name, const std::string& delim = ":", bool validate = true) {
```
With `table_name` and `index_name` also here it's still not immediately clear, but I'm close to being unreasonable on this. So let's go your way.

nyh commented:
CI was hung so rebased to get it running again.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/) | :white_check_mark: | thor |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/)<br> :small_blue_diamond: alternator<br> :small_blue_diamond: alternator/test_gsi_updatetable<br> :small_blue_diamond: alternator/test_lsi<br> :small_blue_diamond: alternator/test_streams<br> :small_blue_diamond: alternator/test_table | :white_check_mark: | spider4.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/)<br> :small_blue_diamond: alternator<br> :small_blue_diamond: alternator/test_gsi_updatetable<br> :small_blue_diamond: alternator/test_lsi<br> :small_blue_diamond: alternator/test_streams<br> :small_blue_diamond: alternator/test_table | :white_check_mark: | spider1.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/) | :white_check_mark: | spider5.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17991/)<br> :small_blue_diamond: alternator<br> :small_blue_diamond: alternator/test_gsi_updatetable<br> :small_blue_diamond: alternator/test_lsi<br> :small_blue_diamond: alternator/test_streams<br> :small_blue_diamond: alternator/test_table | :white_check_mark: | spider5.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5770/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8964/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17048/) | :white_check_mark: |  |

#### Build Details:
- Duration: 1 hr 41 min
